src/gpu/locks.rs

```diff
@@ -14,31 +14,107 @@ use crate::gpu::{CpuGpuMultiexpKernel, GpuName};
 #[derive(Debug)]
-pub struct GPULock(File);
-impl GPULock {
-    pub fn lock() -> GPULock {
-        let gpu_lock_file = tmp_path(GPU_LOCK_NAME);
-        debug!("Acquiring GPU lock at {:?} ...", &gpu_lock_file);
-        let f = File::create(&gpu_lock_file)
-            .unwrap_or_else(|_| panic!("Cannot create GPU lock file at {:?}", &gpu_lock_file));
-        f.lock_exclusive().unwrap();
+pub struct GPULock<'a>(Vec<LockInfo<'a>>);
+
+impl GPULock<'_> {
+    pub fn lock() -> Self {
+        if let Ok(val) = std::env::var("BELLPERSON_GPUS_PER_LOCK") {
+            match val.parse::<usize>() {
+                Ok(val) if val > 0 => {
+                    let devices = Device::all();
+                    info!(
+                        "BELLPERSON_GPUS_PER_LOCK == {}, try lock {}/{} gpus",
+                        val,
+                        val,
+                        devices.len(),
+                    );
+
+                    let mut locks = Vec::new();
+                    for (index, device) in devices.iter().enumerate() {
+                        let uid = device.unique_id();
+                        let path = tmp_path(GPU_LOCK_NAME, Some(uid));
+                        debug!("Acquiring GPU lock {}/{} at {:?} ...", index, val, &path);
+                        let file = File::create(&path).unwrap_or_else(|_| {
+                            panic!("Cannot create GPU {:?} lock file at {:?}", uid, &path)
+                        });
+                        if file.try_lock_exclusive().is_err() {
+                            continue;
+                        }
+                        debug!("GPU lock acquired at {:?}", path);
+                        locks.push(LockInfo {
+                            file: Some(file),
+                            path: Some(path),
+                            devices: vec![device],
+                        });
+                        if locks.len() >= val {
+                            break;
+                        }
+                    }
+
+                    return GPULock(locks);
+                }
+                Ok(val) if val == 0 => {
+                    info!("BELLPERSON_GPUS_PER_LOCK == 0, free to use gpus");
+                    return GPULock(vec![LockInfo {
+                        file: None,
+                        path: None,
+                        devices: Device::all(),
+                    }]);
+                }
+                _ => warn!("BELLPERSON_GPUS_PER_LOCK parse fail, use all gpus"),
+            };
+        }
+
+        info!("BELLPERSON_GPUS_PER_LOCK fallback to single lock mode");
+
+        // Fallback to create single lock
+        let path = tmp_path(GPU_LOCK_NAME, None);
+        debug!("Acquiring GPU lock at {:?} ...", &path);
+        let file = File::create(&path).unwrap_or_else(|_| {
+            panic!("Cannot create GPU lock file at {:?}", &path);
+        });
+        file.lock_exclusive().unwrap();
         debug!("GPU lock acquired!");
```

Comment:
Nitpick: this could be made consistent with the log message from line 64: `debug!("GPU lock acquired at {:?}", path);`.